### PR TITLE
cache reporter

### PR DESCRIPTION
**Pull request by lucamrgs**
Adding documentation and logic for a caching component. It interfaces execution reporting functions, database, and can serve the reporting API.

### CODE REVIEW

**Comment by lucamrgs** [comment]
We can discuss interfaces and components naming, and functionalities.

File: test/unittest/reporters/downstream_reporter/cache_test.go [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+		ExecutionId: executionId0,
+		PlaybookId:  "test",
+		StepResults: map[string]report.StepResult{},
+		Status:      report.Ongoing,
+		Started:     timeNow,
+		Ended:       time.Time{},
+	}
+	expectedExecutions := []string{"6ba7b810-9dad-11d1-80b4-00c04fd430c0"}
+
+	_ = cacheReporter.ReportWorkflowStart(executionId0, playbook)
```
Comment (MaartendeKruijf) [inline test/unittest/reporters/downstream_reporter/cache_test.go:94]: assert error is nil

File: test/unittest/reporters/downstream_reporter/cache_test.go [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+		"6ba7b810-9dad-11d1-80b4-00c04fd430c4",
+		"6ba7b810-9dad-11d1-80b4-00c04fd430c5",
+		"6ba7b810-9dad-11d1-80b4-00c04fd430c6",
+		"6ba7b810-9dad-11d1-80b4-00c04fd430c7",
+		"6ba7b810-9dad-11d1-80b4-00c04fd430c8",
+		"6ba7b810-9dad-11d1-80b4-00c04fd430c9",
+		"6ba7b810-9dad-11d1-80b4-00c04fd430ca",
+	}
+
+	_ = cacheReporter.ReportWorkflowStart(executionId0, playbook)
```
Comment (MaartendeKruijf) [inline test/unittest/reporters/downstream_reporter/cache_test.go:209]: assert error is nil

File: test/unittest/reporters/downstream_reporter/cache_test.go [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+	executionId0, _ := uuid.Parse("6ba7b810-9dad-11d1-80b4-00c04fd430c0")
+
+	layout := "2006-01-02T15:04:05.000Z"
+	str := "2014-11-12T11:45:26.371Z"
+	timeNow, _ := time.Parse(layout, str)
+	mock_time.On("Now").Return(timeNow)
+
+	expectedExecutions := []string{"6ba7b810-9dad-11d1-80b4-00c04fd430c0"}
+
+	_ = cacheReporter.ReportWorkflowStart(executionId0, playbook)
```
Comment (MaartendeKruijf) [inline test/unittest/reporters/downstream_reporter/cache_test.go:297]: assert error is nil

File: test/unittest/reporters/downstream_reporter/cache_test.go [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+	assert.Equal(t, expectedExecutionEntry.ExecutionId, exec.ExecutionId)
+	assert.Equal(t, expectedExecutionEntry.PlaybookId, exec.PlaybookId)
+	assert.Equal(t, expectedExecutionEntry.StepResults, exec.StepResults)
+	assert.Equal(t, expectedExecutionEntry.Status, exec.Status)
+	assert.Equal(t, exec.Ended, expectedExecutionEntry.Ended)
+	assert.Equal(t, err, nil)
+	mock_time.AssertExpectations(t)
+}
+
+func TestReportStepStart(t *testing.T) {
```
Comment (MaartendeKruijf) [inline test/unittest/reporters/downstream_reporter/cache_test.go:320]: combine in one test

File: test/unittest/reporters/downstream_reporter/cache_test.go [hunk 5]
```diff
@@ -0,0 +1,664 @@
...
+
+		Workflow: map[string]cacao.Step{step1.ID: step1, end.ID: end},
+	}
+	executionId0, _ := uuid.Parse("6ba7b810-9dad-11d1-80b4-00c04fd430c0")
+	layout := "2006-01-02T15:04:05.000Z"
+	str := "2014-11-12T11:45:26.371Z"
+	timeNow, _ := time.Parse(layout, str)
+	mock_time.On("Now").Return(timeNow)
+
+	_ = cacheReporter.ReportWorkflowStart(executionId0, playbook)
```
Comment (MaartendeKruijf) [inline test/unittest/reporters/downstream_reporter/cache_test.go:572]: assert error

File: test/unittest/reporters/reporter/reporter_test.go [hunk 6]
```diff
@@ -98,11 +98,82 @@ func TestReportWorkflow(t *testing.T) {
...
+		Workflow: map[string]cacao.Step{step1.ID: step1, end.ID: end},
+	}
+
+	executionId, _ := uuid.Parse("6ba7b810-9dad-11d1-80b4-00c04fd430c8")
+
+	mock_ds_reporter.On("ReportWorkflowEnd", executionId, playbook, nil).Return(nil)
+	reporter.ReportWorkflowEnd(executionId, playbook, nil)
+
+	mock_ds_reporter.AssertExpectations(t)
 }
```
Comment (MaartendeKruijf) [inline test/unittest/reporters/reporter/reporter_test.go:175]: add a test to assert more then one reporter is called (add two downstream reporter mocks and assert they are called)

Comment (MaartendeKruijf) [reply test/unittest/reporters/reporter/reporter_test.go:175]: The is more of an addition to this test, we should have done that with the reporter pull request but lets add it now

**Review by MaartendeKruijf — CHANGES_REQUESTED** [review]
Looks goed only a few minor changes needed great work! @lucamrgs

File: internal/guid/guid.go [hunk 7]
```diff
@@ -9,6 +9,7 @@ IGuid allows one to inject the uuid property and have it be available by mocking
 */
 type IGuid interface {
```
[the remark this answers is not in the document]
Comment (lucamrgs) [reply internal/guid/guid.go:10]: I tested again and we don't. It was a left-over of a previous definition of some reporting functions. I will remove it in the next push.

**Review by MaartendeKruijf — APPROVED** [review]

**Pull request merged**